Site Explorer Panel updates;

aachavarria opened the pull request:
PR includes https://github.com/craftercms/craftercms/issues/4246

aachavarria committed:
update SiteExplorer.tsx: changes on fetchSidebarConfig() validation;
GlobalNav.tsx: adding fetchSidebarConfig() dispatch;

aachavarria committed:
PathNavigator.ts Reducer and Epic;

aachavarria committed:
connecting pathNavigator state to Widget.tsx;

aachavarria committed:
adding localstorage support to pathNavigator.ts;

aachavarria committed:
[4246] GlobalNav menu doesn't work if there are no sites

rart committed:
Format SiteExplorer.tsx

rart commented on ui/app/src/services/configuration.ts:
```diff
@@ -314,10 +314,14 @@ export function getSidebarConfig(site: string): Observable<SidebarConfigItem[]>
-      const cleanDoc = fromString(`<?xml version="1.0" encoding="UTF-8"?><root></root>`);
-      cleanDoc.documentElement.append(...items);
-      return asArray<SidebarConfigItem>(toJS(cleanDoc).root.modulehook).filter(Boolean);
+      if (rawXML) {
```
When did you find this returning something nullish?

aachavarria replied:
When there are not sites, the default values are 'editorial, headless, empty' but when it tries to get the sidebarConfig, it returns null

rart commented on ui/app/src/state/epics/pathNavigator.ts:
```diff
@@ -0,0 +1,48 @@
+      withLatestFrom(state$),
+      mergeMap(([{ type, payload: { id, path } }, state]) => {
+          if (pathNavigatorInit.type === type && localStorage.getItem('pathNavigator')) {
+            return NEVER;
```
Why listen to _Init_ if you don't want to do anything with it?

aachavarria replied:
because pathNavigatorInit.type also needs to call getChildrenByPath, but only when there are not localStorage.getItem('pathNavigator').

rart commented on ui/app/src/state/reducers/pathNavigator.ts:
```diff
@@ -0,0 +1,178 @@
+const reducer = createReducer<LookupTable<WidgetState>>(
+  {},
+  {
+    [pathNavigatorInit.type]: (state, { payload: { props, id } }) => {
```
Pluck the relevant props and send them as direct payload properties

rart commented on ui/app/src/components/Navigation/PathNavigator/Widget.tsx:
```diff
@@ -14,7 +14,7 @@
  */
 
-import React, { Fragment, useCallback, useReducer, useState } from 'react';
+import React, { Fragment, useEffect, useState } from 'react';
```
Heads up: merge coming your way due to [this PR](https://github.com/craftercms/studio-ui/pull/1868) getting merged

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/site-explorer-panel

 Conflicts:
	ui/app/src/components/Navigation/PathNavigator/Widget.tsx
	ui/app/src/state/epics/root.ts
	ui/app/src/state/reducers/root.ts

aachavarria committed:
Fixing Merge Conflicts;
PR Review;

aachavarria committed:
Merge remote-tracking branch 'aaron/feature/site-explorer-panel' into feature/site-explorer-panel

aachavarria committed:
Using PathNavigatorUpdate reducer to update the widget state with the localstorage

rart commented on ui/app/src/state/reducers/pathNavigator.ts:
```diff
@@ -0,0 +1,182 @@
+          items: nextItems,
+          count: response.length
+        };
+        localStorage.setItem(`craftercms.pathNavigator.${id}`, JSON.stringify(widgetState));
```
Reducers should not access local storage. Epics or the component may. 

Don't store data (items). Only store state necessary to restore prior state.

rart commented on ui/app/src/utils/string.ts:
```diff
@@ -50,6 +50,10 @@ export function dasherize(str: string) {
   return str.replace(/_/g, '-');
 }
 
+export function trimerize(str: string) {
```
Rename to removeSpaces

aachavarria committed:
adding pathRestore Ability to Widget.tsx;

aachavarria committed:
creating conditional localStorage recover;

rart committed:
Update GlobalNav.tsx

rart commented on ui/app/src/components/Navigation/PathNavigator/Widget.tsx:
```diff
@@ -508,24 +401,6 @@ export default function (props: WidgetProps) {
-    [dispatch, site, state]
-  );
-
   const [collapsed, setCollapsed] = useState(false);
```
Please add expanded/collapsed as one of the redux state props so we can restore that piece of the state too.

aachavarria committed:
Adding collapsed to localStorage

aachavarria committed:
Merge remote-tracking branch 'aaron/feature/site-explorer-panel' into feature/site-explorer-panel

aachavarria committed:
deleting pathNavigatorFetchPath;
creating epic to save state to localStorage;

aachavarria committed:
fixing PathNavigator Search Styles;
Changes on SearchBar.tsx to support classes;
Adding Refresh button to path Navigator;
adding empty state to ContextMenu.tsx;

rart commented on ui/app/src/state/reducers/pathNavigator.ts:
```diff
@@ -0,0 +1,221 @@
+
+export const pathNavigatorFetchPathFailed = createAction('PATH_NAVIGATOR_FETCH_PATH_FAILED');
+
+export const pathNavigatorSetKeyword = createAction<PayloadWithId<{ keyword: string }>>('PATH_NAVIGATOR_SET_KEYWORD');
```
Move to their own file

aachavarria replied:
done.

rart committed:
Update pathNavigator.ts

aachavarria committed:
Moving actions to its own file

sumerjabri commented:
@aachavarria please fix conflict.

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/site-explorer-panel

 Conflicts:
	ui/app/src/components/Navigation/PathNavigator/Widget.tsx

aachavarria committed:
[3908] ContentTypeEditor: Path outside /site used as "repository path" of datasource provides no warning and breaks model

https://github.com/craftercms/craftercms/issues/3908

aachavarria commented:
@sumerjabri done.
